Tests/SonarScanner.MSBuild.Tasks.IntegrationTests/TargetsTests/RazorTargetTests.cs

```diff
@@ -131,6 +190,62 @@ public void Razor_ExcludedProject_PreserveRazorCompilationErrorLog()
+        [Description("Checks the targets are executed in the expected order")]
+        public void TargetExecutionOrderForRazor()
+        {
+            // Arrange
+            var rootInputFolder = TestUtils.CreateTestSpecificFolderWithSubPaths(TestContext, "Inputs");
+            var rootOutputFolder = TestUtils.CreateTestSpecificFolderWithSubPaths(TestContext, "Outputs");
+            TestUtils.CreateEmptyFile(TestUtils.CreateTestSpecificFolderWithSubPaths(TestContext), "Index.cshtml");
+
+            // We need to set the CodeAnalyisRuleSet property if we want ResolveCodeAnalysisRuleSet
+            // to be executed. See test bug https://github.com/SonarSource/sonar-scanner-msbuild/issues/776
+            var dummyQpRulesetPath = TestUtils.CreateValidEmptyRuleset(rootInputFolder, "dummyQp");
+
+            var projectSnippet = $@"
+<PropertyGroup>
+  <SonarQubeTempPath>{rootInputFolder}</SonarQubeTempPath>
+  <SonarQubeOutputPath>{rootInputFolder}</SonarQubeOutputPath>
+  <SonarQubeConfigPath>{rootOutputFolder}</SonarQubeConfigPath>
+  <CodeAnalysisRuleSet>{dummyQpRulesetPath}</CodeAnalysisRuleSet>
+  <ImportMicrosoftCSharpTargets>false</ImportMicrosoftCSharpTargets>
+  <TargetFramework>net5</TargetFramework>
+  <!-- Prevent references resolution -->
+  <DesignTimeBuild>true</DesignTimeBuild>
+  <GenerateDependencyFile>false</GenerateDependencyFile>
+  <GenerateRuntimeConfigurationFiles>false</GenerateRuntimeConfigurationFiles>
+</PropertyGroup>
+";
+            var txtFilePath = CreateProjectFile(null, projectSnippet, string.Empty);
+            var csprojFilePath = txtFilePath + ".csproj";
+            File.WriteAllText(csprojFilePath, File.ReadAllText(txtFilePath).Replace("<Project ", @"<Project Sdk=""Microsoft.NET.Sdk.Web"" "));
+
+            // Act
+            var result = BuildRunner.BuildTargets(TestContext, csprojFilePath, TargetConstants.DefaultBuild);
+
+            // Assert
+            // Checks that should succeed irrespective of the MSBuild version
+            result.AssertTargetSucceeded(TargetConstants.DefaultBuild);
+            result.AssertTargetExecuted(TargetConstants.OverrideRoslynAnalysis);
+
+            result.AssertExpectedTargetOrdering(
+                TargetConstants.SonarResolveReferences,
+                TargetConstants.SonarOverrideRunAnalyzers,
+                TargetConstants.BeforeCompile,
+                TargetConstants.ResolveCodeAnalysisRuleSet,
+                TargetConstants.SonarCategoriseProject,
+                TargetConstants.OverrideRoslynAnalysis,
+                TargetConstants.SetRoslynAnalysisProperties,
+                TargetConstants.CoreCompile,
+                TargetConstants.SonarPrepareRazorCodeAnalysis,
+                TargetConstants.RazorCoreCompile,
+                TargetConstants.SonarFinishRazorCodeAnalysis,
+                TargetConstants.DefaultBuild,
+                TargetConstants.SetRoslynResults,
+                TargetConstants.SonarWriteProjectData);
```

Comment:
We should actually have this order:
```suggestion
                TargetConstants.CoreCompile,
                TargetConstants.SetRoslynResults,
                TargetConstants.SonarWriteProjectData
                TargetConstants.SonarPrepareRazorCodeAnalysis,
                TargetConstants.RazorCoreCompile,
                TargetConstants.SonarFinishRazorCodeAnalysis, // This could be after build as well
                TargetConstants.DefaultBuild);
```